### PR TITLE
Naga's Contributions

### PR DESCRIPTION
**Pull request by NagaKanaparthy**

### CODE REVIEW

File: cop4610/makefile [hunk 1]
```diff
@@ -0,0 +1,2 @@
+p1: main.c
+  gcc main.c -o p1
```
Comment (JonathanFrias) [inline cop4610/makefile]: This is the type of stuff I was afraid of. There is no tab character (\t) on the second line, here causing the makefile to not work correctly.

File: cop4610/main.c [hunk 2]
```diff
@@ -92,39 +93,46 @@ int main(int argc, char* argv[]) {
...
     }
   }
 
-  for(i = 0; i < atoi(argv[2]); i++) {
+  for(i = 0; i < consumers; i++) {
     if((consumerChildren[i] = fork()) == 0) {
       // child
-      int j ;
-      for(j = 0; j < producers*numItems; j++) {
+      for(int j = 0; j < minPerConsumer; j++) {
```
Comment (JonathanFrias) [inline cop4610/main.c:113]: Really? You converted the code to use c99 and didn't support it in the makefile

File: cop4610/main.c [hunk 3]
```diff
@@ -92,39 +93,46 @@ int main(int argc, char* argv[]) {
...
+        }
+      }
       exit(0);
     }
   }
-
-  for(i = 0 ; i < atoi(argv[1]); i++) {
-    waitpid(producerChildren[i], &exitCode, 0);
+  for(i = 0; i < producers; i++) {
+	  //printf("\nexited %d of %d producers\n", i+1,producers);
```
Comment (JonathanFrias) [inline cop4610/main.c:125]: no tabs!

File: cop4610/main.c [hunk 4]
```diff
@@ -92,39 +93,46 @@ int main(int argc, char* argv[]) {
...
+	  waitpid(producerChildren[i], &exitCode, 0);
```
Comment (JonathanFrias) [inline cop4610/main.c:126]: no tabs!

File: cop4610/main.c [hunk 5]
```diff
@@ -92,39 +93,46 @@ int main(int argc, char* argv[]) {
   int producers = atoi(argv[1]);
   int consumers = atoi(argv[2]);
   int numItems = atoi(argv[3]);
+  int minPerConsumer = (producers*numItems)/consumers;
+  int numExtraItems = (producers*numItems)%consumers;
```
Comment (JonathanFrias) [inline cop4610/main.c:97]: :+1:

File: cop4610/main.c [hunk 6]
```diff
@@ -92,39 +93,46 @@ int main(int argc, char* argv[]) {
...
-  for(i = 0 ; i < atoi(argv[1]); i++) {
-    waitpid(producerChildren[i], &exitCode, 0);
+  for(i = 0; i < producers; i++) {
+	  //printf("\nexited %d of %d producers\n", i+1,producers);
+	  waitpid(producerChildren[i], &exitCode, 0);
   }
-  for(i = 0 ; i < atoi(argv[2]); i++) {
-    waitpid(consumerChildren[i], &exitCode, 0);
+  for(i = 0; i < consumers; i++) {
+	  //printf("\nexited %d of %d consumers\n", i+1,consumers);
```
Comment (JonathanFrias) [inline cop4610/main.c:129]: no tabs

File: cop4610/main.c [hunk 7]
```diff
@@ -156,9 +164,8 @@ int isQueueEmpty() {
  */
 void synchronizedAccess(void (*function)(), bool randomAccessLock, bool produce) {
   if(produce) {
-    semIncrBy(semId, PRODUCER, -1);
-    semIncrBy(semId, CRITICAL, -1);
-
+	semIncrBy(semId, PRODUCER, -1);
+	semIncrBy(semId, CRITICAL, -1);
```
Comment (JonathanFrias) [inline cop4610/main.c:168]: no tabs.

File: cop4610/main.c [hunk 8]
```diff
@@ -73,6 +73,7 @@ void setup() {
 
 int main(int argc, char* argv[]) {
   shmctl(shared_mem_id, IPC_RMID, 0);
+  semctl(semId, IPC_RMID, 0);
```
Comment (JonathanFrias) [inline cop4610/main.c:76]: :+1:

File: cop4610/makefile (hunk 1 above)
Comment (JonathanFrias) [inline cop4610/makefile]: also doesn't compile with c99 support, but uses c99 features.

File: cop4610/Documentation [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+In addition, we have methods to control inc and dec the semaphores.
+They are some assert methods that we used to double check our logic.
+the consume and produce function represent the child processes that form the producers and consumers.
+
+Difficulties:
+We had issues understanding loading the queue plus the COUNTERS into shared memory but moving the pointer was the fix
+Another stupid issue we should have caught was that one consumer was eating all the items. this was do to one of our for loops
+statements being incorrect.
+the semaphore stuff was pretty easy thanks to all the support files given to us by Jim Littlton.
```
Comment (JonathanFrias) [inline cop4610/Documentation]: We also had difficulties contributing before 4 hours before due date!

File: cop4610/main.c [hunk 10]
```diff
@@ -92,39 +93,46 @@ int main(int argc, char* argv[]) {
...
+  for(i = 0; i < consumers; i++) {
     if((consumerChildren[i] = fork()) == 0) {
       // child
-      int j ;
-      for(j = 0; j < producers*numItems; j++) {
+      for(int j = 0; j < minPerConsumer; j++) {
         synchronizedAccess(&consume, true, false);
       }
+      if(i == consumers-1){
+	for(int j = 0; j < numExtraItems; j++) {
```
Comment (JonathanFrias) [inline cop4610/main.c:117]: no c99

**Comment by JonathanFrias** [comment]
Fixes some bugs but also creates new ones.

File: cop4610/main.c (hunk 4 above)
Comment (NagaKanaparthy) [inline cop4610/main.c:126]: My bad edited on text editor

File: cop4610/main.c (hunk 2 above)
Comment (NagaKanaparthy) [inline cop4610/main.c:113]: the osprey c compiler auto uses c11 to compile but i should have included that in the makefile

File: cop4610/makefile [hunk 11]
```diff
@@ -0,0 +1,2 @@
+p1: main.c
+	gcc main.c -std=c11 -o p1
```
Comment (JonathanFrias) [inline cop4610/makefile]:
```
cc1: error: unrecognized command line option "-std=c11"
make: *** [p1] Error 1
```

This still doesn't work. It's so much easier to just not support c99 for this project.

That means:

```
for(int i = 0; ;) { }
```

becomes 

```
int i;
for(i = 0 ; ;) { }
```

File: cop4610/main.c (hunk 2 above)
Comment (JonathanFrias) [inline cop4610/main.c:113]: No. look at this on osprey.

```
-bash-4.1$ hostname
osprey.unfcsd.unf.edu
-bash-4.1$ make
gcc main.c -std=c11 -o p1
cc1: error: unrecognized command line option "-std=c11"
make: *** [p1] Error 1
```

File: cop4610/makefile [hunk 12]
```diff
@@ -0,0 +1,2 @@
+p1: main.c
+	gcc main.c -std=c99 -o p1
```
Comment (JonathanFrias) [inline cop4610/makefile:2]: :+1:

File: cop4610/main.c (hunk 2 above)
Comment (JonathanFrias) [inline cop4610/main.c:113]: Okay this is fine now because makefile was updated.

**Pull request merged**